File: backend/src/sandbox/api/tool/grep.py
"""Internal implementation for the public sandbox grep verb."""

from __future__ import annotations

from audit.base import AuditSink
from sandbox.api.tool._daemon_payload import daemon_request_identity
from sandbox.api.tool._operation_audit import run_audited_operation
from sandbox.api.tool._result_projection import grep_result_from_daemon_response
from sandbox.api.timeouts import GREP_TIMEOUT_S
from sandbox.api.transport import DAEMON_OP_GREP, DaemonSandboxTransport, SandboxTransport
from sandbox._shared.models import GrepRequest, GrepResult


async def grep(
    sandbox_id: str,
    request: GrepRequest,
    *,
    audit_sink: AuditSink | None = None,
    transport: SandboxTransport | None = None,
) -> GrepResult:
    """Regex-scan workspace file contents under the sandbox's leased snapshot."""
    daemon_transport = transport or DaemonSandboxTransport()

    async def _call() -> GrepResult:
        payload = daemon_request_identity(request) | {
            "pattern": request.pattern,
            "output_mode": request.output_mode,
            "offset": request.offset,
            "case_insensitive": request.case_insensitive,
            "line_numbers": request.line_numbers,
            "multiline": request.multiline,
        }
        if request.path is not None:
            payload["path"] = request.path
        if request.glob_filter is not None:
            payload["glob_filter"] = request.glob_filter
        if request.head_limit is not None:
            payload["head_limit"] = request.head_limit
        response = await daemon_transport.call(
            sandbox_id,
            DAEMON_OP_GREP,
            payload,
            timeout=GREP_TIMEOUT_S,
        )
        return grep_result_from_daemon_response(response)

    return await run_audited_operation(
        audit_sink=audit_sink,
        sandbox_id=sandbox_id,
        operation="grep",
        caller=request.caller,
        payload={
            "pattern": request.pattern,
            "path": request.path or "",
            "output_mode": request.output_mode,
        },
        call=_call,
    )


__all__ = ["grep"]

File: agent-core/crates/eos-sandbox-api/src/tool_api/grep.rs
//! Pure `grep` helper: build payload → call transport → parse envelope.

use eos_types::SandboxId;
use serde_json::Value;

use crate::error::SandboxApiError;
use crate::models::{GrepRequest, GrepResult};
use crate::ops::DaemonOp;
use crate::timeouts::GREP_TIMEOUT_S;
use crate::tool_api::parse::{daemon_request_identity_fields, parse_grep_result};
use crate::transport::SandboxTransport;

/// Regex-scan workspace file contents under the sandbox's leased snapshot.
pub async fn grep(
    transport: &dyn SandboxTransport,
    sandbox_id: &SandboxId,
    request: &GrepRequest,
) -> Result<GrepResult, SandboxApiError> {
    let mut payload = daemon_request_identity_fields(&request.base);
    payload.insert("pattern".to_owned(), Value::String(request.pattern.clone()));
    payload.insert(
        "output_mode".to_owned(),
        Value::String(request.output_mode.clone()),
    );
    payload.insert("offset".to_owned(), Value::from(request.offset));
    payload.insert(
        "case_insensitive".to_owned(),
        Value::Bool(request.case_insensitive),
    );
    payload.insert("line_numbers".to_owned(), Value::Bool(request.line_numbers));
    payload.insert("multiline".to_owned(), Value::Bool(request.multiline));
    if let Some(path) = &request.path {
        payload.insert("path".to_owned(), Value::String(path.clone()));
    }
    if let Some(glob_filter) = &request.glob_filter {
        payload.insert("glob_filter".to_owned(), Value::String(glob_filter.clone()));
    }
    if let Some(head_limit) = request.head_limit {
        payload.insert("head_limit".to_owned(), Value::from(head_limit));
    }
    let response = transport
        .call(sandbox_id, DaemonOp::Grep, payload, GREP_TIMEOUT_S)
        .await?;
    parse_grep_result(&response)
}
